src/graph/immutable_graph.cc

```diff
@@ -0,0 +1,975 @@
+    size_t row_start = indptr[oldvid];
+    size_t row_len = indptr[oldvid + 1] - indptr[oldvid];
+    def_check.CollectOnRow(&indices[row_start], &edge_ids[row_start], row_len,
+                           &sub_csr->indices, &orig_edge_ids);
+    sub_csr->indptr[i + 1] = sub_csr->indices.size();
+  }
+
+  // Store the non-zeros in a subgraph with edge attributes of new edge ids.
+  sub_csr->edge_ids.resize(sub_csr->indices.size());
+  for (int64_t i = 0; i < sub_csr->edge_ids.size(); i++)
+    sub_csr->edge_ids[i] = i;
+
+  IdArray rst_eids = IdArray::Empty({static_cast<int64_t>(orig_edge_ids.size())},
+                                    DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+  dgl_id_t* eid_data = static_cast<dgl_id_t*>(rst_eids->data);
+  std::copy(orig_edge_ids.begin(), orig_edge_ids.end(), eid_data);
+
+  return std::pair<ImmutableGraph::CSR::Ptr, IdArray>(sub_csr, rst_eids);
+}
+
+ImmutableGraph::CSR::Ptr ImmutableGraph::CSR::FromEdges(std::vector<Edge> *edges,
+                                                        int sort_on, int64_t num_nodes) {
+  CHECK(sort_on == 0 || sort_on == 1) << "we must sort on the first or the second vector";
+  int other_end = sort_on == 1 ? 0 : 1;
+  // TODO(zhengda) we should sort in parallel.
+  struct compare {
+    int sort_on;
+    int other_end;
+    compare(int sort_on, int other_end) {
+      this->sort_on = sort_on;
+      this->other_end = other_end;
+    }
+    bool operator()(const Edge &e1, const Edge &e2) {
+      if (e1.end_points[sort_on] == e2.end_points[sort_on])
+        return e1.end_points[other_end] < e2.end_points[other_end];
+      else
+        return e1.end_points[sort_on] < e2.end_points[sort_on];
+    }
+  };
+  std::sort(edges->begin(), edges->end(), compare(sort_on, other_end));
+  auto t = std::make_shared<CSR>(0, 0);
+  t->indices.resize(edges->size());
+  t->edge_ids.resize(edges->size());
+  for (size_t i = 0; i < edges->size(); i++) {
+    t->indices[i] = edges->at(i).end_points[other_end];
+    CHECK(t->indices[i] < num_nodes);
+    t->edge_ids[i] = edges->at(i).edge_id;
+    dgl_id_t vid = edges->at(i).end_points[sort_on];
+    CHECK(vid < num_nodes);
+    while (vid > 0 && t->indptr.size() <= static_cast<size_t>(vid))
+      t->indptr.push_back(i);
+    CHECK(t->indptr.size() == vid + 1);
+  }
+  while (t->indptr.size() < num_nodes + 1)
+    t->indptr.push_back(edges->size());
+  CHECK(t->indptr.size() == num_nodes + 1);
+  return t;
+}
+
+ImmutableGraph::CSR::Ptr ImmutableGraph::CSR::Transpose() const {
+  std::vector<Edge> edges(NumEdges());
+  for (size_t i = 0; i < NumVertices(); i++) {
+    const dgl_id_t *indices_begin = &indices[indptr[i]];
+    const dgl_id_t *eid_begin = &edge_ids[indptr[i]];
+    for (size_t j = 0; j < GetDegree(i); j++) {
+      Edge e;
+      e.end_points[0] = i;
+      e.end_points[1] = indices_begin[j];
+      e.edge_id = eid_begin[j];
+      edges[indptr[i] + j] = e;
+    }
+  }
+  return FromEdges(&edges, 1, NumVertices());
```

Comment:
Magic number `1`? Use a bool?

Reply:
If it is only 1 or 0, change the type of the argument to bool and use names like `sort_on_src_or_dst` and doc it.